multicluster/service-mirror/cluster_watcher.go

```diff
@@ -815,3 +937,307 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateEndpoints(ctx context.Con
 
 	return nil
 }
+
+func (rcsw *RemoteClusterServiceWatcher) createOrUpdateHeadlessEndpoints(ctx context.Context, remoteEp *corev1.Endpoints) error {
+	// Get local version of the Endpoints object and diff
+	localEpName := rcsw.mirroredResourceName(remoteEp.Name)
+	localEndpoints, err := rcsw.localAPIClient.Endpoint().Lister().Endpoints(remoteEp.Namespace).Get(localEpName)
+	localEp := localEndpoints.DeepCopy()
+
+	// If Endpoints update on the remote cluster, and the mirrorerd version of
+	// the Endpoints does not exist then it means we need to create it (and
+	// all of its nested clusterIP services).
+	if err != nil {
+		if !kerrors.IsNotFound(err) {
+			return err
+		}
+
+		if err := rcsw.handleHeadlessEndpointsCreated(ctx, remoteEp); err != nil {
+			rcsw.log.Debugf("failed to create headless mirrors for Endpoints %s/%s: %v", remoteEp.Namespace, remoteEp.Name, err)
+			return err
+		}
+
+		return nil
+	}
+
+	remoteService, err := rcsw.remoteAPIClient.Svc().Lister().Services(remoteEp.Namespace).Get(remoteEp.Name)
+	if err != nil {
+		rcsw.log.Debugf("failed to update headless endpoints for %s/%s:%v", remoteEp.Namespace, remoteEp.Name, err)
```

Comment:
Hope I didn't go overboard here but I think you're right -- in the spirit of being more descriptive I added on to both the Debug statement and the returned error.

I wanted to change the error returned to also give more context -- these errors are usually logged by the event processor loop; it can be really easy to lose context and figure out where things went wrong. I think the more descriptive error will help with that. Lmk what ya think.